---
title: 'figure 1b: Comparing ChIP-nexus and ChIP-seq of Oct4 and Sox2'
author: "Melanie (1028-02098-001-001)"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

# Introduction

We want to plot ChIP-nexus and ChIP-seq peaks based on canonical Oct/Sox, Sox, and Klf4 motifs for ChIP-seq and ChIP-nexus data to compare resolution quality.

# Computational Setup

```{r, warning=F, message=F}
#Standard packages
source("https://bioconductor.org/biocLite.R")
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools); library(data.table)
library(BSgenome.Mmusculus.UCSC.mm10); library(plyr); library(viridis); library(dplyr)

#KNITR Options
setwd("/n/projects/mw2098/publications/2019_bpnet/")
options(knitr.figure_dir="15-fig1b-oct4-sox2-comparison", width=16, height=20)

#Lab sources
source("rscripts/granges_common.r")
source("rscripts/metapeak_common.r")
source("rscripts/knitr_common.r")
source("rscripts/caching.r")
source("rscripts/multiplot.R")

#Specific packages
library(grid)
library(lattice)
library(gtable)
library(latticeExtra)
library(gridExtra)
library(Hmisc)
```

# Define data paths

ChIP samples were normalized WRT reads per million (RPM).

```{r define samples}
samples.list<-list(nexus = list(oct4 = list(pos = "data/bw/nexus/mesc_oct4_nexus_combined_filtered_normalized_positive.bw",
                                            neg = "data/bw/nexus/mesc_oct4_nexus_combined_filtered_normalized_negative.bw"),
                                sox2 = list(pos = "data/bw/nexus/mesc_sox2_nexus_combined_filtered_normalized_positive.bw",
                                            neg = "data/bw/nexus/mesc_sox2_nexus_combined_filtered_normalized_negative.bw")),
                   chip = list (oct4 = list(rep1 = "data/bw/seq/mesc_oct4_seq_1_normalized.bw",
                                            rep2 = "data/bw/seq/mesc_oct4_seq_2_normalized.bw"),
                                sox2 = list(rep1 = "data/bw/seq/mesc_sox2_seq_1_normalized.bw",
                                            rep2 = "data/bw/seq/mesc_sox2_seq_2_normalized.bw")))
```

# Locate regions

Given:

OCT4SOX2 motif: YYWTTGT  
SOX2 motif: YYWTTGTNATGCAAA  

These motifs are derived from previous de novo motif analysis in the lab and previous literature defining these canonical motifs (TherMos: Estimating proteinâ€“DNA binding energies from in vivo binding profiles, 2013). 

```{r motif locations, eval=F}
sox2_motif <- DNAString("RGRACAAWRGV") %>% 
  vmatchPattern(., BSgenome.Mmusculus.UCSC.mm10, max.mismatch =0, fixed = "subject") 
sox2oct4_motif <- DNAString("SYWTTGTNATGCAAA") %>% 
  vmatchPattern(., BSgenome.Mmusculus.UCSC.mm10, max.mismatch =0, fixed = "subject") 
sox2oct4_motif_1mm <- DNAString("SYWTTGTNATGCAAA") %>% 
  vmatchPattern(., BSgenome.Mmusculus.UCSC.mm10, max.mismatch =1, fixed = "subject") 
klf4_motif <- DNAString("WGGGYGKGGY") %>% 
  vmatchPattern(., BSgenome.Mmusculus.UCSC.mm10, max.mismatch =0, fixed = "subject")
```

# Calculate counts

In order to find the strongest peaks for the representative plots, we will count the reads occurring at a 100bp region for each sample.

```{r calculate counts, eval=F}
sox2_motif$sox2_signals<-abs(regionSums(regions = resize(sox2_motif, 100, fix="center"), cvg = samples.list$nexus$sox2$pos)) + 
  abs(regionSums(regions = resize(sox2_motif, 100, fix="center"), cvg = samples.list$nexus$sox2$neg))
sox2oct4_motif$oct4_signals<-abs(regionSums(regions = resize(sox2oct4_motif, 100, fix="center"), cvg = samples.list$nexus$oct4$pos)) + 
  abs(regionSums(regions = resize(sox2oct4_motif, 100, fix="center"), cvg = samples.list$nexus$oct4$neg))
sox2oct4_motif_1mm$oct4_signals<-abs(regionSums(regions = resize(sox2oct4_motif_1mm, 100, fix="center"), cvg = samples.list$nexus$oct4$pos)) + 
  abs(regionSums(regions = resize(sox2oct4_motif, 100, fix="center"), cvg = samples.list$nexus$oct4$neg))

saveRDS(sox2_motif, "data/rdata/motif_locations/sox2_motif_no_mm.rds")
saveRDS(sox2oct4_motif, "data/rdata/motif_locations/sox2oct4_motif_0_mm.rds")
saveRDS(sox2oct4_motif_1mm, "data/rdata/motif_locations/sox2oct4_motif_1_mm.rds")
```

# Order the data 

Order the data based on counts and select top 500 peaks for each motif. Filter Sox2 instances from Sox2-Oct4 motifs.

```{r order}
#Import data
sox2_motif<-readRDS("data/rdata/motif_locations/sox2_motif_no_mm.rds") %>% keepStandardChromosomes(pruning.mode = "coarse", species = "Mus_musculus")
sox2oct4_motif<-readRDS("data/rdata/motif_locations/sox2oct4_motif_1_mm.rds") %>% keepStandardChromosomes(pruning.mode = "coarse", species = "Mus_musculus")

#Keep overlaps off
sox2_motif<-sox2_motif[-findOverlaps(resize(sox2_motif, 500, "center"), resize(sox2oct4_motif, 500, "center"))@from]

#Select top peaks
ntop<-500
ntop_corrected_for_oct4<-ntop

if(ntop>length(sox2oct4_motif)){ntop_corrected_for_oct4<-length(sox2oct4_motif)}

sox2_motif<-sox2_motif[order(sox2_motif$sox2_signals, decreasing=T)][1:(ntop)] 
sox2oct4_motif<-sox2oct4_motif[order(sox2oct4_motif$oct4_signals, decreasing=T)][1:ntop_corrected_for_oct4]

#Define motifs
motif.list<-list(sox2 = sox2_motif,
                 oct4 = sox2oct4_motif)
```

# Get metapeaks

Obtain metapeak profiles as dataframes.

```{r get metapeaks}
#nexus
nexus_metapeaks<-lapply(c("sox2", "oct4"), function(x){
  lapply(c("sox2", "oct4"), function(y){
      message(x, y)
      peaks<-exo_metapeak(gr = resize(motif.list[[x]], 1, "center"), sample = samples.list$nexus[[y]], upstream = 200, downstream = 200) 
      peaks$fac<-y
      peaks$motif<-x
      peaks[,c(1,2,3,6,7)]
  }) %>% rbindlist
}) %>% rbindlist

#seq
seq_metapeaks<-lapply(c("sox2", "oct4"), function(x){
  lapply(c("sox2", "oct4"), function(y){
      peaks<-exo_metapeak(gr = resize(motif.list[[x]], 1, "center"), sample = list(pos=samples.list$chip[[y]]$rep1, neg=samples.list$chip[[y]]$rep2), upstream = 200, downstream = 200)
      peaks<-peaks %>% group_by(tss_distance) %>% dplyr::summarize(reads=sum(abs(reads))/2)
      peaks$strand<-"*"
      peaks$fac<-y
      peaks$motif<-x
      peaks
  }) %>% rbindlist
}) %>% rbindlist
```

# Plot metapeaks

Plot metapeaks together.

```{r fig1b-oct4-sox2-comparison}
#Normalize and distinguish
nexus_metapeaks$norm_reads<-nexus_metapeaks$reads/max(nexus_metapeaks$reads)
nexus_metapeaks$type<-"nexus"
seq_metapeaks$norm_reads<-seq_metapeaks$reads/max(seq_metapeaks$reads)
seq_metapeaks$type<-"seq"

combined_metapeaks<-rbind(seq_metapeaks, nexus_metapeaks)
combined_metapeaks$fac<-factor(combined_metapeaks$fac, levels=c("sox2", "oct4"), labels=c("Sox2 ChIP", "Oct4 ChIP"))
combined_metapeaks$motif<-factor(combined_metapeaks$motif, levels=c("sox2", "oct4"), labels=c("Sox2 Motif", "Sox2-Oct4 Motif"))
combined_metapeaks$strand<-factor(combined_metapeaks$strand, levels=c("*", "-", "+"))
combined_metapeaks$type<-factor(combined_metapeaks$type, levels=c("seq", "nexus"))
combined_metapeaks$tss_distance<-combined_metapeaks$tss_distance %>% as.character %>% as.numeric

#Plot
showcase<-ggplot(subset(combined_metapeaks, type=="seq"), aes(x=tss_distance, y=norm_reads))+
  geom_area(aes(fill=type), position="identity")+
  geom_line(data=subset(combined_metapeaks, type=="nexus"), aes(color=fac))+
  facet_grid(fac ~ motif)+
  scale_alpha_manual(values = c(.5, 1), guide=F)+
  scale_fill_manual(values = c("grey50"))+
  scale_color_manual(values = c("#3A3C97", "#9F1D20"))+
  xlab("Distance from motif start")+
  ylab("Normalized Signal")+
  theme_classic()+
  theme(text=element_text(size=20), strip.background = element_blank())
print(showcase)
#ggsave("figures/showcase/overlay_combined/showcase_of_motifs.pdf", showcase, height=12, width=12)
# 
# showcase_line<-ggplot(subset(combined_metapeaks, type=="seq"), aes(x=tss_distance, y=norm_reads))+
#   geom_area(aes(fill=type), position="identity")+
#   geom_line(data=subset(combined_metapeaks, type=="nexus"), aes(color=fac, group=strand))+
#   facet_grid(fac ~ motif)+
#   scale_alpha_manual(values = c(.5, 1), guide=F)+
#   scale_fill_manual(values = c("grey50"))+
#   scale_color_manual(values = c("#3A3C97", "#9F1D20"))+
#   xlab("Distance from motif start")+
#   ylab("Normalized Signal")+
#   theme_classic()+
#   theme(text=element_text(size=20), strip.background = element_blank())
# print(showcase_line)
# 
# ggsave("figures/showcase/overlay_combined/showcase_of_motifs_line_plot.pdf", showcase_line, height=12, width=12)
```

# Session Information

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```













