---
title: "19-sfig4a-nanog-sox: Investigation of Nanog-Sox Heterodimer"
author: "Melanie (1028-02098-001-001)"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

# Introduction

It has been previously proposed that a Nanog-Sox heterodimer is a possible motif for optimal recruitment of Nanog and Sox2: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3746198/. However, when running BPnet we have not found evidence of this occurring. In order to address this discrepancy in results, we want to do the following:

1. Obtain a PWM from the Nanog-Sox heterodimer from Figure 7C of the paper listed above. Use this PWM to scan the mm10 genome to obtain genomic instances of this heterodimer. Compare these genomic instances to the Nanog-specific and Sox2 motif in terms of frequency and overlap.
2. Query these SELEX sequences in BPnet to see how BPnet treats these regions.

# Computational Setup

```{r, warning=F, message=F}
#Standard packages
source("https://bioconductor.org/biocLite.R")
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools); library(data.table)
library(BSgenome.Mmusculus.UCSC.mm10); library(plyr); library(viridis); library(dplyr)

#KNITR Options
setwd("/n/projects/mw2098/publications/2019_bpnet/")
options(knitr.figure_dir="19-sfig4a-nanog-sox", width=16, height=20)

#Lab sources
source("rscripts/granges_common.r")
source("rscripts/knitr_common.r")
source("rscripts/caching.r")
source("rscripts/multiplot.R")
source("rscripts/metapeak_functions.R")
source("rscripts/metapeak_common.r")

#Specific packages
library(stringr)
library(rBLAST)
library(motifStack)
library(grid)
library(gridExtra)
```

# Define data paths

```{r define data}
#.TXT files with sequences from Gagliardi et al: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3746198/, Figure 7C
CATT_seq_filepath <- "data/txt/gagliardi_selex_sequences_aligned_to_CATT.txt"

#.TXT files with same sequences as link above, but with manual realignment to the TGA, not the CATT component of the sequences.
TGA_seq_filepath <- "data/txt/gagliardi_selex_sequences_julias_alignment.txt"

# List of ChIP-nexus samples that were combined across replicates, blacklisted according to mm10, and normalized to reads per million (RPM). 
bw_actual_list<-list(oct4 = list(pos = "data/bw/nexus/mesc_oct4_nexus_combined_filtered_normalized_positive.bw",
                                 neg = "data/bw/nexus/mesc_oct4_nexus_combined_filtered_normalized_negative.bw"),
                     sox2 = list(pos = "data/bw/nexus/mesc_sox2_nexus_combined_filtered_normalized_positive.bw",
                                  neg = "data/bw/nexus/mesc_sox2_nexus_combined_filtered_normalized_negative.bw"),
                     nanog=list(pos="data/bw/nexus/mesc_nanog_nexus_combined_filtered_normalized_positive.bw",
                                neg="data/bw/nexus/mesc_nanog_nexus_combined_filtered_normalized_negative.bw"),
                     klf4=list(pos="data/bw/nexus/mesc_klf4_nexus_combined_filtered_normalized_positive.bw",
                               neg="data/bw/nexus/mesc_klf4_nexus_combined_filtered_normalized_negative.bw"))

#List of BPNet predictions of ChIP-nexus samples listed above.
bw_predicted_list<-list(oct4=list(pos="data/bw/predictions/Oct4.preds.pos.bw",
                               neg="data/bw/predictions/Oct4.preds.neg.bw"),
                     sox2=list(pos="data/bw/predictions/Sox2.preds.pos.bw",
                               neg="data/bw/predictions/Sox2.preds.neg.bw"),
                     nanog=list(pos="data/bw/predictions/Nanog.preds.pos.bw",
                               neg="data/bw/predictions/Nanog.preds.neg.bw"),
                     klf4=list(pos="data/bw/predictions/Klf4.preds.pos.bw",
                               neg="data/bw/predictions/Klf4.preds.neg.bw"))

#List of BPNet importance scores of ChIP-nexus profiles listed above.
bw_imp_list<-list(oct4="data/bw/importance/Oct4.importance.profile.bw",
                  sox2="data/bw/importance/Sox2.importance.profile.bw",
                  nanog="data/bw/importance/Nanog.importance.profile.bw",
                  klf4="data/bw/importance/Klf4.importance.profile.bw")

#BED file with coordinates of each seqlet location returned by TF-MoDISco.
seqlet_coordinates_filepath <- "data/bed/seqlet_coordinates.bed"

#BED file with coordinates of each peak location trained with BPNet
peak_coordinates_filepath <- "data/bed/peak_coordinates.bed"
```

# Read in the Gagliardi et al. sequences

Thanks to Julia, teh sequences from Gagliardi et al (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3746198/) are typed out in a text file! We padded the sequences with N's to align them to CATT (CATT_seq) and maintain Julia's alignment of sequences (TGA_seq). Both will be checked for matches.

```{r import sequences}
CATT_seq<-read.table(CATT_seq_filepath, stringsAsFactors = F)$V1 %>% as(Class="DNAStringSet") %>% subseq(start=10, width=15) #As consistent with Gagliardi et al.
TGA_seq<-read.table(TGA_seq_filepath, stringsAsFactors = F)$V1 %>% as(Class="DNAStringSet") %>% subseq(start=16, width=14)
```

# Create Nanog-Sox2 PWM

```{r construct PWM}
CATT_PWM<-CATT_seq %>% PWM
TGA_PWM<-TGA_seq %>% PWM
```

# Match PWM across genome

```{r PWM match, eval=F}
CATT_PWM_matches<-matchPWM(pwm = CATT_PWM, subject = BSgenome.Mmusculus.UCSC.mm10, min.score = "90%")
saveRDS(CATT_PWM_matches, "data/rdata/nanog-sox/gagliardi_nanog_sox_pwm_matches_CATT_aligned.granges.rds")

TGA_PWM_matches<-matchPWM(pwm = TGA_PWM, subject = BSgenome.Mmusculus.UCSC.mm10, min.score = "90%")
saveRDS(TGA_PWM_matches, "data/rdata/nanog-sox/gagliardi_nanog_sox_pwm_matches_TGA_aligned.granges.rds")
```

```{r read PWM matches}
CATT_PWM_matches<-readRDS("data/rdata/nanog-sox/gagliardi_nanog_sox_pwm_matches_CATT_aligned.granges.rds")
TGA_PWM_matches<-readRDS("data/rdata/nanog-sox/gagliardi_nanog_sox_pwm_matches_TGA_aligned.granges.rds")
```

# Generate a graphic motif for the Nanog-Sox2 

As reported by Gagliardi, generate a graphical representation of the Nanog-Sox2 motif.

```{r sfig4a-nanog-sox-motif-logo}
#Plot pos motif
CATT_PWM_matches$seq<-getSeq(BSgenome.Mmusculus.UCSC.mm10, CATT_PWM_matches)
CATT_PWM_matches.pwm<-PWM(CATT_PWM_matches$seq, type="prob")
CATT_PWM_matches_motif <- new("pcm", mat=as.matrix(CATT_PWM_matches.pwm), name="Sox2-Nanog as reported by Gagliardi et. al.")
plot(CATT_PWM_matches_motif)

#Save to PDF
pdf("19-sfig4a-nanog-sox/sfig4a-nanog-sox-motif-logo.pdf", width = 7.8, height=3)
plot(CATT_PWM_matches_motif)
dev.off()
```

# Are the PWMs matched across TF MODISCO regions?

```{r pwm matches across TF MODISCO}
seqlets<-import(seqlet_coordinates_filepath)

#CATT
CATT_overlaps<-findOverlaps(CATT_PWM_matches, seqlets)
CATT_overlaps@to %>% unique %>% length

seqlets_with_CATT_overlaps<-seqlets[unique(CATT_overlaps@to)]
seqlets_with_CATT_overlaps$name %>% table

#TGA
TGA_overlaps<-findOverlaps(TGA_PWM_matches, seqlets)
TGA_overlaps@to %>% unique %>% length

seqlets_with_TGA_overlaps<-seqlets[unique(TGA_overlaps@to)]
seqlets_with_TGA_overlaps$name %>% table
```

#Does BPnet think these regions are of high importance?

Are the CATT N-S heterodimer matches highly important? In order to subset from a smaller set of 201K sequence matches, we will overlap them with relevant peaks that were used to train BPnet.

Once done, we will check 5 sets: OSKN and CATT motif regions for 3 things: importance, predicted signal, and actual signal.

*NOTE: We will be using the DeepLift run, not the published run. The results are similar enough that this will suffice.*

### Collect setup

```{r BPnet importance for CATT regions}
#Collect motif regions
peaks<-import(peak_coordinates_filepath)
CATT_peak_overlaps<-subsetByOverlaps(CATT_PWM_matches, peaks)  %>% granges #4834 matches
CATT_peak_overlaps$name<-"CATT_PWM"
CATT_peak_overlaps$score<-0
all_motif_regions<-c(subset(seqlets, name=="Oct4_0" | name=="Sox2_1" | name=="Nanog_1" | name=="Klf4_0"), CATT_peak_overlaps)
```

### Compute signals

Generate predicted signals, actual signals, and importance scores. 

```{r compute signals, eval=F}
all_motif_signals<-mclapply(unique(all_motif_regions$name), function(x){
  
  message(x)
  gr<-subset(all_motif_regions, name==x) #subset the GRanges
  
  #actual signals
  actual_temp<-lapply(bw_actual_list, function(a){
    sig<-regionMeans(regions = resize(gr, 100, "center"), a$pos) + abs(regionMeans(regions = resize(gr, 100, "center"), a$neg))
    return(data.frame(mean_signal=mean(sig), med_signal=median(sig), sd_signal=sd(sig), type="actual_signal"))
  }) %>% rbindlist(idcol = "fac")
  
  #predicted signals
  predicted_temp<-lapply(bw_predicted_list, function(b){
    sig<-regionMeans(regions = resize(gr, 100, "center"), b$pos) + abs(regionMeans(regions = resize(gr, 100, "center"), b$neg))
    return(data.frame(mean_signal=mean(sig), med_signal=median(sig), sd_signal=sd(sig), type="predicted_signal"))
  }) %>% rbindlist(idcol = "fac")  
  
  #importance
  imp_temp<-lapply(bw_imp_list, function(c){
    sig<-regionMeans(regions = gr, c)
    return(data.frame(mean_signal=mean(sig), med_signal=median(sig), sd_signal=sd(sig), type="importance"))
  }) %>% rbindlist(idcol = "fac")    
  
  combined_temp<-rbind(actual_temp, predicted_temp, imp_temp) #Combine the signals for plotting
  combined_temp$motif<-x
  return(combined_temp)
  
}, mc.cores=5) %>% rbindlist

saveRDS(all_motif_signals, "data/rdata/nanog-sox/signal_comparisons_of_CATT.rds")
```

### Plot comparisons

```{r sfig4a-nanog-sox-motif-heatmaps}
all_motif_signals<-readRDS("data/rdata/nanog-sox/signal_comparisons_of_CATT.rds")
all_motif_signals$motif<-factor(all_motif_signals$motif, levels=rev(c("Oct4_0", "Sox2_1", "Nanog_1", "Klf4_0", "CATT_PWM")), labels=rev(c("Oct4-Sox2", "Sox2", "Nanog", "Klf4", "Putative Motif")))
all_motif_signals$fac<-factor(all_motif_signals$fac, levels=c("oct4", "sox2", "nanog", "klf4"), labels=c("Oct4", "Sox2", "Nanog", "Klf4"))

#Actual signal
sig<-ggplot(subset(all_motif_signals, type=="actual_signal"), aes(fac, motif, fill=med_signal))+
  geom_tile(color="black")+
  scale_fill_gradient(low = "white", high="#48723E", name="RPM")+
  ggtitle("Actual signals across motifs")+
  theme_classic()+
  theme(text=element_text(size=16))
print(sig)

#Predicted signal
pred<-ggplot(subset(all_motif_signals, type=="predicted_signal"), aes(fac, motif, fill=med_signal))+
  geom_tile(color="black")+
  scale_fill_gradient(low = "white", high="#4A5A76", name="RPM")+
  ggtitle("Predicted signals across motifs")+
  theme_classic()+
  theme(text=element_text(size=16))
print(pred)

#Importance
imp<-ggplot(subset(all_motif_signals, type=="importance"), aes(fac, motif, fill=med_signal))+
  geom_tile(color="black")+
  scale_fill_gradient(low = "white", high="#47255F", name="IMP")+
  ggtitle("Importance across motifs")+
  theme_classic()+
  theme(text=element_text(size=16))
print(imp)

#Write as a combined plots
grid.newpage()
pdf("19-sfig4a-nanog-sox/sfig4a-nanog-sox-heatmaps.pdf", width = 7, height = 12) # Open a new pdf file
grid.draw(rbind(rbind(ggplotGrob(sig), ggplotGrob(pred), size = "first"), ggplotGrob(imp), size = "first"))
dev.off() # Close the file
```

No signal, importance, or predicted signals are high across these N-S heterodimer PWM matches.

# Session Information

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```






















