## TODO

- [ ] merge in all the notes
- [ ] what do we need to execute before we run the ipynbs?
- [ ] Is `'dfi_subset.annotated.parq'` really required:
  - Used in `./7.4-modisco-bpnet-comparison.ipynb:3009:    "dfi = pd.read_parquet(models_dir / 'dfi_subset.annotated.parq', engine='fastparquet')`

## Pre-process the data

Run the following commands:

1. `snakemake -s Snakefile.fimo.smk` - Scan the genome with PFM's generated by TF modisco
1. `snakemake -s Snakefile.dfi_subset.smk` - Generate the `dfi_subset` files.
1. ? `python generate_annotated_dfi.py` - annotate the `dfi_subset` with other features
1. Execute `calibrate-count-prediction.ipynb` to obtain `{model_dir}/calibrated_seqmodel.pkl`

## Notebook index

- [01-model-evaluation.test.ipynb](01-model-evaluation.test.ipynb)
  - Figure 1 D,E
  - Figure S1 A,B
- [02-model-evaluation.hyper-param.ipynb](02-model-evaluation.hyper-param.ipynb)
  - Figure 1 F
  - Figure S1 C-E
- [03-representative-motifs.ipynb](03-representative-motifs.ipynb)
  - Figure S3
  - Figure 4 A-C
- [04-motif-instance-stats.ipynb](04-motif-instance-stats.ipynb)
  - Figure 2D,E
- [05-locus-plots.ipynb](05-locus-plots.ipynb)
  - Figure 2B
  - Figure S2 A,B,C,D
- [06-long-motifs.ipynb](06-long-motifs.ipynb)
  - Figure 3 C,D
  - Figure 3 B,E,F
- [07-motif-interactions.synthetic.ipynb](07--motif-interactions.synthetic.ipynb)
  - Figure 5 C
  - Figure S6 B,C
  - Supplementary movie 1
- [08-motif-interactions.genomic.ipynb](08-motif-interactions.genomic.ipynb)
  - Figure 5 D,E
  - Figure S7A,C,D
- [09-spacing.motif-instances.ipynb](09-spacing.motif-instances.ipynb)
  - Figure 5F
  - Figure 6A,H,I
  - Figure S6B
- [10-spacing.contrib-periodicity.ipynb](10-spacing.contrib-periodicity.ipynb)
  - Figure 6 D,E,G,J
- [11-method-comparison-pwm-chexmix.ipynb](11-method-comparison-pwm-chexmix.ipynb)
  - Figure 2F
  - Figure 6 A,B,C
  - Figure ST1 A,B,C,D
- [12-method-comparison-binary,chipseq.ipynb](12-method-comparison-binary,chipseq.ipynb)
  - Figure ST2 A,B,C,D
  - Figure ST3 E,F
- [13-bpnet-chipseq-locus.ipynb](13-bpnet-chipseq-locus.ipynb)
  - Figure ST3 A,D
- [14-bpnet-chipseq-evaluation.ipynb](14-bpnet-chipseq-evaluation.ipynb)  
  - Figure ST3 B,C
- [15-fig1b-oct4-sox2-comparison](15-fig1b-oct4-sox2-comparison.Rmd)
  - Figure 1B
- [16-fig3g_sfig4b-strict-motif-spacing](16-fig3g_sfig4b-strict-motif-spacing.Rmd)
  - Figure 3G, S4B
- [17-fig4d-nanog-variants](17-fig4d-nanog-variants.Rmd)
  - Figure 4D
- [18-fig4ef_sfig5ef-tfiiic-bbox](18-fig4ef_sfig5ef-tfiiic-bbox.Rmd)
  - Figure 4E-F, S5E-F
- [19-sfig4a-nanog-sox](19-sfig4a-nanog-sox.Rmd)
  - Figure S4A
- [20-sfig5abcd-motif-validation](20-sfig5abcd-motif-validation.Rmd)
  - Figure S5A-D
- [21-sfig8-nanog-periodicity](21-sfig8-nanog-periodicity.py)
  - Figure S8
